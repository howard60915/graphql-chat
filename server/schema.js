const schema = `
input RegisterByEmailMutationInput {
  email: String!
  password: String!
}

type RegisterByEmailMutationPayload {
  status: ResponseStatus
  user: User
}

input UpdateUserMutationInput {
  userId: String!
  email: String
}

type UpdateUserMutationPayload {
  status: ResponseStatus
  user: User
}

input DelUserMutationInput {
  userId: String!
}

type DelUserMutationPayload {
  status: ResponseStatus
  user: User
}

input AddConversationMutationInput {
  name: String
}

type AddConversationMutationPayload {
  status: ResponseStatus
  conversation: Conversation
}

input AddMessageMutationInput {
  conversationId: String!
  content: String!
}

type AddMessageMutationPayload {
  status: ResponseStatus
  message: Message
}

type Mutation {
  registerByEmail(input: RegisterByEmailMutationInput!): RegisterByEmailMutationPayload
  updateUser(input: UpdateUserMutationInput!): UpdateUserMutationPayload
  delUser(input: DelUserMutationInput!): DelUserMutationPayload
  addConversation(input: AddConversationMutationInput!): AddConversationMutationPayload
  addMessage(input: AddMessageMutationInput!): AddMessageMutationPayload
}

type Query {
  users: [User]
  conversations: [Conversation]
  messages: [Message]
}

type Subscription {
  onNewMessage: Message
}

enum ResponseStatus {
  failed
  ok
}

type User {
  id: String
  email: String
  isAdmin: Boolean
}

type Conversation {
  id: String
  name: String
}

type Message {
  id: String
  user: User
  conversation: Conversation
  content: String
}
`;

export default schema;
