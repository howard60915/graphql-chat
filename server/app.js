import graphqlServer from "./graphql";

const PORT = process.env.NODE_PORT || process.env.PORT || 8080;

const options = {
  port: PORT,
  endpoint: "/graphql",
  playground: "/playground"
};

graphqlServer.start(options, ({ port }) => {
  // eslint-disable-next-line no-console
  console.log(`> Ready on http://localhost:${port}`);
});
