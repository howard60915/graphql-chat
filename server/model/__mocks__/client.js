import { Client } from "pg";
import { mock } from "graphql-shortcake";
import knex from "knex";
import config from "../../../knexfile";

const { POSTGRES_HOST } = process.env;

const client = jest.requireActual("../client").default;

const db = knex(config);

const mockClient = mock(client);

mockClient.reset = async name => {
  const pgClient = new Client({
    host: POSTGRES_HOST || "127.0.0.1",
    user: "postgres",
    password: null,
    database: "postgres"
  });
  await pgClient.connect();
  await pgClient.query(`DROP DATABASE IF EXISTS ${name}`);
  await pgClient.query(`CREATE DATABASE ${name}`);
  await pgClient.end();
  await db.migrate.latest();
};

export default mockClient;
