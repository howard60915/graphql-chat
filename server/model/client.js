import { Client } from "graphql-shortcake";
import config from "../../knexfile";

export default new Client({ connectionString: config.connection });
