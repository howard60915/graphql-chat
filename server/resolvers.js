import UsersQuery from "./query/UsersQuery";
import RegisterByEmailMutation from "./mutation/RegisterByEmailMutation";
import UpdateUserMutation from "./mutation/UpdateUserMutation";
import DelUserMutation from "./mutation/DelUserMutation";
import AddMessageMutation from "./mutation/AddMessageMutation";

const Query = {
  users: UsersQuery
};

const Mutation = {
  registerByEmail: RegisterByEmailMutation,
  updateUser: UpdateUserMutation,
  delUser: DelUserMutation,
  addMessage: AddMessageMutation
};

export default { Query, Mutation };
