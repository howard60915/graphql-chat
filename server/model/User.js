import { GlobalModel, Hash } from "graphql-shortcake";
import client from "./client";
import Conversation from "./Conversation";

export default class User extends GlobalModel {
  static client = client;

  static hasOperator = true;

  static operator = () => User;

  static tableName = "user";

  static columns = {
    email: { type: String },
    password: { type: Hash, paths: ["archive"] },
    isAdmin: { type: Boolean, paths: ["archive"] }
  };

  async checkWritability(user, ThrowError) {
    const result = this.id === user.id || user.isAdmin;
    if (!result) throw new ThrowError();
    return result;
  }

  async haveConversationWith(user) {
    if (!user) throw new Error("Talker required");

    const conversation = new Conversation();
    conversation.name = `${this.email}, ${user.email}`;
    await conversation.save(this);

    return conversation;
  }
}
