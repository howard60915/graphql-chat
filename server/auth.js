import { NotFoundError } from "graphql-schema-sushi";
import { AuthenticationError } from "apollo-server-errors";
import { JsonWebToken } from "nextjs-apollo/server";
import User from "./model/User";
import Certificate from "./model/Certificate";

const onSignIn = async data => {
  const { email, password, correlationId } = data;

  const { fetcher } = User;
  const user = await fetcher.where("email", email).fetch(NotFoundError);

  if (!(password && user.password.verify(password))) {
    throw new AuthenticationError("AuthenticationError");
  }

  return { user: user.previous, correlationId };
};

const onSignOut = () => ({});

const onRenew = async data => {
  const { user, correlationId } = data;

  const renewUser = await User.load(user.id, NotFoundError);

  return { user: renewUser.previous, correlationId };
};

const toUser = ({ user }) => {
  if (!user.id) throw new AuthenticationError();
  return User.forge(user);
};

const jwt = new JsonWebToken(Certificate, {
  onSignIn,
  onSignOut,
  onRenew,
  toUser
});

export default jwt;
