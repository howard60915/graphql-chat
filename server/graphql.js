import { GraphQLServer } from "graphql-yoga";
import auth from "./auth";

const server = new GraphQLServer({
  typeDefs: `
    type Query {
      hello: String
    }
  `,
  resolvers: {
    Query: {
      hello: () => "hello"
    }
  }
});

auth.applyMiddleware(server.express);

export default server;
