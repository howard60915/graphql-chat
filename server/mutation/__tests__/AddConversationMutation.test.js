import client from "../../model/client";
import User from "../../model/User";
import server, { mutation } from "../../__tests__";

const { addConversation } = mutation;

describe("AddConversationMutation", () => {
  const userId = new User("00000000-0000-0000-0000-000000000015").id;
  const input = { userId };

  afterAll(() => new Promise(() => server.close()));

  it("successfully add conversation", async () => {
    client.mockReturnValueOnce([
      {
        id: "00000000-0000-0000-0000-000000000015",
        email: "world@chat.com"
      }
    ]);
    client.mockReturnValueOnce([
      {
        id: "00000000-0000-0000-0000-000000000020",
        name: "hello@chat.com, world@chat.com"
      }
    ]);
    client.mockReturnValueOnce([
      {
        id: "00000000-0000-0000-0000-000000000020",
        conversation_id: "00000000-0000-0000-0000-000000000020",
        user_id: "00000000-0000-0000-0000-000000000010"
      }
    ]);
    client.mockReturnValueOnce([
      {
        id: "00000000-0000-0000-0000-000000000020",
        conversation_id: "00000000-0000-0000-0000-000000000020",
        user_id: "00000000-0000-0000-0000-000000000015"
      }
    ]);
    const { errors, data } = await addConversation.user({ input });
    expect(errors).toBeUndefined();
    expect(client).toMatchSnapshot();
    expect(client).toHaveBeenCalledTimes(4);
    expect(data).toEqual({
      addConversation: {
        status: "ok",
        conversation: {
          name: "hello@chat.com, world@chat.com"
        }
      }
    });
  });

  it("when talker not found", async () => {
    client.mockReturnValueOnce([]);
    const { errors } = await addConversation.user({ input });
    expect(errors[0].message).toBe("not found");
  });
});
