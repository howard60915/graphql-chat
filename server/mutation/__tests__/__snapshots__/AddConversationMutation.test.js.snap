// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddConversationMutation successfully add conversation 1`] = `
Array [
  "SELECT * FROM \\"user\\" WHERE \\"user\\".\\"deleted_at\\" IS NULL AND \\"id\\" IN ($1)",
  "INSERT INTO \\"conversation\\" (\\"name\\",\\"created_by\\",\\"updated_by\\",\\"created_at\\",\\"updated_at\\") VALUES ($1,$2,$3,$4,$5) RETURNING *",
  "INSERT INTO \\"user_conversation\\" (\\"user_id\\",\\"conversation_id\\",\\"created_by\\",\\"updated_by\\",\\"created_at\\",\\"updated_at\\") VALUES ($1,$2,$3,$4,$5,$6) RETURNING *",
  "INSERT INTO \\"user_conversation\\" (\\"user_id\\",\\"conversation_id\\",\\"created_by\\",\\"updated_by\\",\\"created_at\\",\\"updated_at\\") VALUES ($1,$2,$3,$4,$5,$6) RETURNING *",
]
`;
